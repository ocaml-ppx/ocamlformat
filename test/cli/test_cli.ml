let pp_banner ppf s =
  let line = String.make (String.length s) '=' in
  Format.fprintf ppf "%s\n%s\n" s line

let pp_status ppf = function
  | Unix.WEXITED n -> Format.fprintf ppf "[%d]" n
  | WSIGNALED _ -> Format.fprintf ppf "WSIGNALED _"
  | WSTOPPED _ -> Format.fprintf ppf "WSTOPPED _"

let main ocamlformat =
  let test ?(stdin = "") ~name ~args () =
    let cmd = Printf.sprintf "%s %s 2>&1" ocamlformat args in
    let proc = Unix.open_process cmd in
    let proc_out, proc_in = proc in
    Stdio.Out_channel.output_string proc_in stdin ;
    Stdio.Out_channel.close proc_in ;
    let out = Stdio.In_channel.input_all proc_out in
    let stat = Unix.close_process proc in
    Format.printf "%aargs: %s\n%s%a\n\n" pp_banner name args out pp_status stat
  in
  let impl_file = Stdio.In_channel.read_all "sample/b.ml" in
  let intf_file = Stdio.In_channel.read_all "sample/a.mli" in
  test ~name:"err_default_several_file" ~args:"sample/a.ml sample/b.ml" () ;
  test ~name:"err_inplace_and_check" ~args:"--inplace --check sample/a.ml" () ;
  test ~name:"err_inplace_and_output"
    ~args:"--inplace --output o.ml sample/a.ml" () ;
  test ~name:"err_no_arg" ~args:"" () ;
  test ~name:"err_output_and_check" ~args:"--output x.ml --check sample/a.ml" () ;
  test ~name:"err_output_several_files"
    ~args:"--output x.ml sample/a.ml sample/b.ml" () ;
  test ~name:"err_stdin_and_file" ~args:"sample/a.ml -" () ;
  test ~name:"err_stdin_and_inplace" ~args:"--inplace -" () ;
  test ~name:"err_stdin_no_kind" ~args:"-" () ;
  test ~name:"stdin_and_impl" ~stdin:impl_file ~args:"--impl -" () ;
  test ~name:"stdin_and_intf" ~stdin:intf_file ~args:"--intf -" () ;
  test ~name:"stdin_and_name" ~stdin:impl_file ~args:"--name a.ml -" () ;
  test ~name:"name_unknown_ext" ~args:"--name b.cpp sample/b.ml" () ;
  test ~name:"err_stdin_name_unknown_ext" ~stdin:impl_file
    ~args:"--name b.cpp -" ()

let () =
  match Sys.argv with
  | [|_; ocamlformat|] -> main ocamlformat
  | _ -> assert false
