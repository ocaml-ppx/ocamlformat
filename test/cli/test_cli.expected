err_default_several_file
========================
args: sample/a.ml sample/b.ml
ocamlformat: Must specify exactly one input file without --inplace or --check
[1]

err_inplace_and_check
=====================
args: --inplace --check sample/a.ml
ocamlformat: Cannot specify --inplace with --check
[1]

err_inplace_and_output
======================
args: --inplace --output o.ml sample/a.ml
ocamlformat: Cannot specify --output with --inplace
[1]

err_no_arg
==========
args: 
ocamlformat: Must specify at least one input file, or `-` for stdin
[1]

err_output_and_check
====================
args: --output x.ml --check sample/a.ml
ocamlformat: Cannot specify --output with --check
[1]

err_output_several_files
========================
args: --output x.ml sample/a.ml sample/b.ml
ocamlformat: Must specify exactly one input file without --inplace or --check
[1]

err_stdin_and_file
==================
args: sample/a.ml -
ocamlformat: Cannot specify stdin together with other inputs
[1]

err_stdin_and_inplace
=====================
args: --inplace -
ocamlformat: Must specify at least one of --name, --impl or --intf when reading from stdin
[1]

err_stdin_no_kind
=================
args: -
ocamlformat: Must specify at least one of --name, --impl or --intf when reading from stdin
[1]

stdin_and_impl
==============
args: --impl -
let () = print_endline A.x
[0]

stdin_and_intf
==============
args: --intf -
val x : string
[0]

stdin_and_name
==============
args: --name a.ml -
let () = print_endline A.x
[0]

name_unknown_ext
================
args: --name b.cpp sample/b.ml
let () = print_endline A.x
[0]

err_stdin_name_unknown_ext
==========================
args: --name b.cpp -
ocamlformat: Must specify at least one of --name, --impl or --intf when reading from stdin
[1]

